public_html/wp-content/mu-plugins/blocks/assets/src/blocks-store.js

```diff
@@ -0,0 +1,199 @@
+
+	apiFetchResult.then(
+		( customEntities ) => {
+			dispatch( WC_BLOCKS_STORE ).setEntities( entityType, customEntities );
+		}
+	).catch( //TODO: Implement retries on HTTP Transport errors.
+		( reason ) => {
+			console.error( entityType, 'Unable to retrieve data from API.', reason );
+		}
+	);
+};
+
+/**
+ * Define actions which can be dispatched by this store.
+ *
+ * @type {Object}
+ */
+const actions = {
+	/**
+	 * Queues fetching from API for a post type.
+	 *
+	 * @param {string} entityType
+	 * @returns {Object}
+	 */
+	fetchEntities( entityType ) {
+		return {
+			type       : 'FETCH_ENTITIES',
+			entityType : entityType,
+		};
+	},
+
+	/**
+	 * Set entities state
+	 *
+	 * @param {string} entityType
+	 * @param {Array}  entities
+	 *
+	 * @return {Object}
+	 */
+	setEntities( entityType, entities ) {
+		return {
+			type       : 'SET_ENTITIES',
+			entityType : entityType,
+			entities   : entities,
+		};
+	},
+};
+
+/**
+ * Defines selectors provided by this store.
+ *
+ * @type {{getPosts(*, *)}}
+ */
+const selectors = {
+
+	/**
+	 * Returns post type from current state. Caches [state, entityType] for quick resolution.
+	 *
+	 * @param {Object} state      Current store state
+	 * @param {string} entityType Name of the entity to get
+	 * @param {Object} args       Additional filter arguments.
+	 */
+	getEntities: createSelector(
+		( state, entityType, args = {} ) => {
+			let results = state[ entityType ];
+
+			if ( ! state.hasOwnProperty( entityType ) ) {
+				// Queue apiFetch for this entity.
+				dispatch( WC_BLOCKS_STORE ).fetchEntities( entityType );
+				return;
+			}
+
+			if ( Array.isArray( args.entityIds ) ) {
+				results = results.filter( ( item ) => -1 !== args.entityIds.indexOf( item.id ) );
+			}
+
+			if ( 'function' === typeof args.filterEntities ) {
+				results = results.filter( args.filterEntities );
+			}
+
+			if ( 'function' === typeof args.orderBy ) {
+				results = args.orderBy( results );
+			}
+
+			return results;
+		},
+		// Return state if entityType is yet initialized to prevent unnecessary selector executions.
+		( state, entityType ) => state[ entityType ] || state
+	),
+};
+
+registerStore(
+	WC_BLOCKS_STORE,
+	{
+		/**
+		 * Reducer for this store.
+		 *
+		 * @param {Object} state
+		 * @param {Object} action
+		 *
+		 * @returns {Object}
+		 */
+		reducer( state = DEFAULT_STATE, action ) {
+			switch ( action.type ) {
+				case 'FETCH_ENTITIES':
+
+					apiFetchEntities(
+						state,
+						action.entityType,
+						API_ARGS.entity.path( action.entityType ),
+						API_ARGS.entity.query
+					);
+
+					break;
+
+				case 'SET_ENTITIES':
+
+					// Changing state reference so that withSelect works.
+					state = { ...state };
+					state[ action.entityType ] = action.entities;
+
+					// Not really needed, but lets do this for correctness.
+					state.loadingEntities = state.loadingEntities.filter( ( item ) => action.entityType !== item );
+
+					break;
+			}
+			return state;
+		},
```

Comment:
So `reducer` is always a single function instead of objects as in the case of `actions` and `selectors`. So generally its written this way (i.e. reducer fully defined, while `actions`, `selectors`, `revolvers` etc are in separate objects if they are complicated. Its just cosmetics, but we can make it a separate function if you feel strongly about this though.

Reply:
Yeah, I feel like the "idiomatic JavaScript" conventions are often really horrible for readability. That could just be my personal bias, but to me this feels like something that would be objectively easier to read as a modular function, and then the registration would be super quick & easy to parse:

```js
registerStore(
	WC_BLOCKS_STORE,
	{ selectors, actions, reducer },
);
```

Reply:
If you and/or @coreymckrill disagree, though, I'm happy to figure out something we're all happy with.

Reply:
I looked at how the `registerStore` method is being implemented in the `core-data` module in Gutenberg, and it defines the reducer outside of the register call. So I moved ours outside of the register call as well: fcb4a35c509ca71b8065f4f3b8f189eeca9594d2